starfish/intensity_table.py

```diff
@@ -291,3 +293,123 @@ def mask_small_features(self, size_threshold):
     def _intensities_from_regions(self, props, reduce_op='max') -> "IntensityTable":
         """turn regions back into intensities by reducing over the labeled area"""
         raise NotImplementedError
+
+    def combine_adjacent_features(
+            self, area_threshold, assume_contiguous=True
+    ) -> Tuple["IntensityTable", List, np.ndarray, np.ndarray, np.ndarray]:
+        """
+
+        Parameters
+        ----------
+        area_threshold
+        assume_contiguous
+
+        Notes
+        -----
+
+        --
+        # This is how a non-contiguous ImageStack creation could work:
+        # construct an empty (z, y, x) image
+        decoded_image = np.zeros(self.attrs['image_shape'], dtype=int)
+
+        # fill it with features
+        coordinates = self.indexes['features'].to_frame()[['z', 'y', 'x']]
+        genes = self.gene_name.values
+
+        for i in np.arange(self.sizes[self.Constants.FEATURES.value]):
+            z, y, x = coordinates.iloc[i]
+            decoded_image[z, y, x] = gene_to_int[genes[i]]
+
+        --
+        # ideally we would be able to define a function that would extract the original
+        # area of the now-merged spot and calculate characteristics of it (max intensity, etc)
+        # across the code. It's signature could look a bit like this:
+        def props_to_intensities(props, label_image, image_stack, reduce_function
+                                 ) -> "IntensityTable":
+            tiles: List[np.ndarray] = []
+            for prop in props:
+                # extract the relevant area of the table and apply the reduce function
+                raise NotImplementedError
+
+        Returns
+        -------
+        IntensityTable :
+            spot intensities
+        List[
+
+        """
+        if not assume_contiguous:
+            raise NotImplementedError
+
+        # None needs to map to zero, non-none needs to map to something else.
+        int_to_gene = dict(
+            zip(range(1, np.iinfo(np.int).max),
+                set(self.gene_name.values) - {'None'}))
+        int_to_gene[0] = 'None'
+        gene_to_int = {v: k for (k, v) in int_to_gene.items()}
+
+        # map genes to ints
+        gene_list = [gene_to_int[g] for g in self.coords[self.Constants.GENE.value].values]
+        gene_array = np.array(gene_list)
+        decoded_image = gene_array.reshape(self.attrs['image_shape'])  # reverse linearization
+
+        # label the image and extract max intensity across each feature
+        label_image = label(decoded_image, connectivity=2)
+        intensities = self.max(dim=[Indices.CH, Indices.HYB])  # todo can use mean here too
```

Comment:
interesting. if i'm reading this right, for each 'pixel' you record it's maximum intensity across C and H. then, as pixel vectors that decode to the same spot are merged into a 'spot', we record the max of these max intensities across pixels in a spot.  I think the max of the max of a bunch of L2 normalized vectors is an odd statistic to report. the max of the mean, however, feels more natural. perhaps we can expose an option here to report the max or the mean that defaults to the mean to be parallel with the GaussianSpotDetector? How do we handle this in the LocalMaxPeakFinder?

Reply:
> interesting. if i'm reading this right, for each 'pixel' you record it's maximum intensity across C and H. then, as pixel vectors that decode to the same spot are merged into a 'spot', we record the max of these max intensities across pixels in a spot.

You read this correctly. 

> perhaps we can expose an option here to report the max or the mean that defaults to the mean to be parallel with the GaussianSpotDetector? How do we handle this in the LocalMaxPeakFinder?

LocalMaxPeakFinder reports the max intensity of the spot, but it's not general to multiplex assays (see discussion in #305). It might be a good idea to hash out exactly what API we want to expose for spot finders.